Added log levels

KeerthanaAP opened the pull request:
Fix https://github.com/ppc64le-cloud/pvsadm/issues/34
As mentioned in the below table added log levels for the existing log messages.


Level | Purpose
-- | --
0 | Info
1 | Debug
2 | Trace level 1
3 | Trace level 2
4 | Function entry and exit

mkumatag commented:
@KeerthanaAP what is the default log level(what level messages will print if user doesn't mention anything)?

KeerthanaAP commented:
> at level messages will print if user doesn't mention anythin

KeerthanaAP commented:
> @KeerthanaAP what is the default log level(what level messages will print if user doesn't mention anything)?

With the current changes, Only warnings and errors will be printed by default. 
By default log level will be set to 0. User has to provide the log level 1 to print the info messages.

mkumatag commented:
> With the current changes, Only warnings and errors will be printed by default.
> By default log level will be set to 0. User has to provide the log level 1 to print the info messages.

Either we need to change the default level to some value and print the meaningful messages with that default level otherwise it won't print most of the messages.

KeerthanaAP commented:
> > With the current changes, Only warnings and errors will be printed by default.
> > By default log level will be set to 0. User has to provide the log level 1 to print the info messages.
> 
> Either we need to change the default level to some value and print the meaningful messages with that default level otherwise it won't print most of the messages.

Sure. Can we move all the log level one down as mentioned below. so that by default all the info messages will be printed.

Level | Purpose
-- | --
0 | Info
1 | Debug
2 | Trace level 1
3 | Trace level 2
4 | Function entry and exit

mkumatag commented:
> Sure. Can we move all the log level one down as mentioned below. so that by default all the info messages will be printed.

sgtm

KeerthanaAP commented:
> > Sure. Can we move all the log level one down as mentioned below. so that by default all the info messages will be printed.
> 
> sgtm

Done.Modified all the log level as mentioned and updated in the description

KeerthanaAP commented:
/retest-required

mkumatag commented:
/cc @kishen-v

kishen-v commented on cmd/image/import/import.go:
```diff
@@ -217,7 +217,7 @@ pvsadm image import -n upstream-core-lon04 -b <BUCKETNAME>  --object rhel-83-100
 				return err
 			}
-			klog.Infof("Retriving image details")
+			klog.V(1).Infof("Retriving image details")
```
```suggestion
			klog.V(1).Infof("Retriving image details")
```
```suggestion
			klog.V(1).Infof("Retrieving image details")
```

KeerthanaAP replied:
Done.

kishen-v commented on cmd/image/qcow2ova/prep/prepare.go:
```diff
@@ -21,7 +21,6 @@ import (
 	"strings"
 
 	"github.com/ppc64le-cloud/pvsadm/pkg/utils"
-
```
We may need to follow the order of packages :
standard packages.
3rd party. 
internal packages.

KeerthanaAP replied:
Modified the order of packages.

kishen-v commented on cmd/image/info/info.go:
```diff
@@ -58,13 +58,13 @@ pvsadm image info rhcos-46-12152021.ova.gz
 				return err
 			}
-			klog.Infof("Extract complete")
+			klog.V(1).Infof("Extract complete")
```
No formatting directives used, we may use `Info`.
```suggestion
			klog.V(1).Info("Extract complete")
```

KeerthanaAP replied:
Done.

kishen-v commented on cmd/image/qcow2ova/get-image.go:
```diff
@@ -40,18 +40,18 @@ func getImage(downloadDir string, srcUrl string, timeout time.Duration) (string,
 			return "", err
 		}
-		klog.Infof("Copy Completed!")
+		klog.V(1).Infof("Copy Completed!")
```
```suggestion
		klog.V(1).Info("Copy Completed!")
```

KeerthanaAP replied:
Done.

kishen-v commented on cmd/image/qcow2ova/get-image.go:
```diff
@@ -69,7 +69,7 @@ func getImage(downloadDir string, srcUrl string, timeout time.Duration) (string,
 			return "", err
 		}
-		klog.Infof("Download Completed!")
+		klog.V(1).Infof("Download Completed!")
```
```suggestion
		klog.V(1).Info("Download Completed!")
```

KeerthanaAP replied:
Done.

kishen-v commented on cmd/image/qcow2ova/qcow2ova.go:
```diff
@@ -215,13 +229,13 @@ Qcow2 images location:
 				return err
 			}
-			klog.Infof("Extract complete")
+			klog.V(1).Infof("Extract complete")
```
```suggestion
			klog.V(1).Info("Extract complete")
```

KeerthanaAP replied:
Done.

kishen-v commented on cmd/root.go:
```diff
@@ -47,7 +47,7 @@ This is a tool built for the Power Systems Virtual Server helps managing and mai
 		if pkg.Options.APIKey == "" {
 			if key := os.Getenv("IBMCLOUD_API_KEY"); key != "" {
-				klog.Infof("Using an API key from IBMCLOUD_API_KEY environment variable")
+				klog.V(1).Infof("Using an API key from IBMCLOUD_API_KEY environment variable")
```
```suggestion
				klog.V(1).Info("Using an API key from IBMCLOUD_API_KEY environment variable")
```

KeerthanaAP replied:
Done.

kishen-v commented on pkg/client/volume/volume.go:
```diff
@@ -55,7 +55,7 @@ func (c *Client) DeleteVolume(id string) error {
 
 func (c *Client) GetAll() (*models.Volumes, error) {
-	klog.Infof("Calling the Power Volumes GetAll Method")
+	klog.V(1).Infof("Calling the Power Volumes GetAll Method")
```
```suggestion
	klog.V(1).Info("Calling the Power Volumes GetAll Method")
```

KeerthanaAP replied:
Done.

kishen-v commented on test/e2e/sync/sync.go:
```diff
@@ -65,7 +65,7 @@ func runSyncCMD(args ...string) (int, string, string) {
 // Create Specifications yaml file
 func createSpecFile(spec []pkg.Spec) error {
-	klog.Infof("STEP: Creating Spec file")
+	klog.V(3).Infof("STEP: Creating Spec file")
```
```suggestion
	klog.V(3).Info("STEP: Creating Spec file")
```

KeerthanaAP replied:
Done.

mkumatag commented:
@kishen-v can you review this again?

kishen-v commented on cmd/dhcp-sync/dhcp-sync.go:
```diff
@@ -184,10 +184,10 @@ var Cmd = &cobra.Command{
-					klog.Info("event:", event)
+					klog.V(2).Info("event:", event)
 					if event.Op&fsnotify.Write == fsnotify.Write {
-						klog.Info("modified file:", event.Name)
-						klog.Info("restarting the dhcpd service")
+						klog.V(2).Info("modified file:", event.Name)
```
Suggested usage of `Infof` over `Info` although the latter achieves the same objective. 
This can be done to keep changes in-line with the rest of the project.

KeerthanaAP replied:
Done.

kishen-v commented on cmd/image/import/import.go:
```diff
@@ -161,7 +161,7 @@ pvsadm image import -n upstream-core-lon04 -b <BUCKETNAME>  --object rhel-83-100
 			} else {
 				// Use the service credential already created
-				klog.Infof("Reading the existing service credential")
+				klog.V(2).Infof("Reading the existing service credential")
```
```suggestion
				klog.V(2).Info("Reading the existing service credential")
```

KeerthanaAP replied:
Done.

kishen-v commented on cmd/image/sync/sync_test.go:
```diff
@@ -79,7 +79,7 @@ func TestGetSpec(t *testing.T) {
 		SpecFileName := file.Name()
 		defer os.Remove(SpecFileName)
-		klog.Infoln("Specfile :", SpecFileName)
+		klog.V(1).Infoln("Specfile :", SpecFileName)
```
Usage of `Infof` preferred.

KeerthanaAP replied:
Done.

kishen-v commented on cmd/image/sync/sync_test.go:
```diff
@@ -79,7 +79,7 @@ func TestGetSpec(t *testing.T) {
 		SpecFileName := file.Name()
 		defer os.Remove(SpecFileName)
-		klog.Infoln("Specfile :", SpecFileName)
+		klog.V(1).Infoln("Specfile :", SpecFileName)
```
`Infof` over `Infoln`

KeerthanaAP replied:
Done.

kishen-v commented on pkg/client/s3client.go:
```diff
@@ -118,7 +118,7 @@ func NewS3Client(c *Client, instanceName, region string) (s3client *S3Client, er
 	result, err := c.S3Session.ListBuckets(nil)
 	if err != nil {
-		klog.Infof("Unable to list buckets, %v\n", err)
+		klog.V(1).Infof("Unable to list buckets, %v\n", err)
```
```suggestion
		klog.V(1).Infof("Unable to list buckets, %v", err)
```
Ref: https://pkg.go.dev/k8s.io/klog#Infof

KeerthanaAP replied:
Done.

kishen-v commented on cmd/image/qcow2ova/prep/loop.go:
```diff
@@ -122,7 +122,7 @@ func Umount(dir string) error {
 				}
 			}
-			klog.Infof(" As '%s' is busy, unmounting it using lazy unmount", dir)
+			klog.V(1).Infof(" As '%s' is busy, unmounting it using lazy unmount", dir)
```
```suggestion
			klog.V(1).Infof("As '%s' is busy, unmounting it using lazy unmount", dir)
```

KeerthanaAP replied:
Done

KeerthanaAP commented:
/retest

KeerthanaAP commented on test/e2e/sync/sync.go:
```diff
@@ -330,7 +330,7 @@ func verifyBucketObjects(tgt pkg.TargetItem, cos string, files []fs.FileInfo, re
 		fileName := f.Name()
 		res := false
-		klog.Infoln("Verifying object", fileName)
+		klog.V(4).Infoln("Verifying object", fileName)
```
yeah, Modified it all other similar logs to V(3)

KeerthanaAP commented on test/e2e/sync/sync.go:
```diff
@@ -164,7 +164,7 @@ func deleteCOSInstance(instanceName string) error {
 // Create S3 bucket in the given region and storage class
 func createBucket(bucketName string, cos string, region string, storageClass string) error {
-	klog.Infof("STEP: Creating Bucket %s in region %s in COS %s storageClass %s", bucketName, region, cos, storageClass)
+	klog.V(3).Infof("STEP: Creating Bucket %s in region %s in COS %s storageClass %s", bucketName, region, cos, storageClass)
```
yes,  Changed it V(4)

KeerthanaAP commented on test/e2e/sync/sync.go:
```diff
@@ -65,7 +65,7 @@ func runSyncCMD(args ...string) (int, string, string) {
 // Create Specifications yaml file
 func createSpecFile(spec []pkg.Spec) error {
-	klog.Infof("STEP: Creating Spec file")
+	klog.V(3).Info("STEP: Creating Spec file")
```
Done.

KeerthanaAP commented:
@kishen-v  Addressed all the comments, could you ptal.

kishen-v commented:
@KeerthanaAP, sure!

kishen-v commented on cmd/image/qcow2ova/qcow2ova.go:
```diff
@@ -159,10 +160,23 @@ Qcow2 images location:
+				return err
+			}
+
+			klog.Infof("Autogenerated OS root password is updated in file :  %s/%s", cwd, passwordfile.Name())
```
Ref: https://pkg.go.dev/path/filepath#Join
```suggestion
			klog.Infof("Autogenerated OS root password is updated in %s", filepath.Join(cwd, passwordfile.Name()))
```

KeerthanaAP replied:
Done.

kishen-v commented on pkg/client/network/network.go:
```diff
@@ -57,7 +57,7 @@ func (c *Client) GetAll() (*models.Networks, error) {
 
 	if err != nil || resp.Payload == nil {
-		klog.Infof("Failed to perform the operation... %v", err)
+		klog.V(1).Infof("Failed to perform the operation... %v", err)
```
```suggestion
		klog.V(1).Errorf("Failed to perform the operation... %v", err)
```

kishen-v commented on pkg/client/s3client.go:
```diff
@@ -118,7 +118,7 @@ func NewS3Client(c *Client, instanceName, region string) (s3client *S3Client, er
 	result, err := c.S3Session.ListBuckets(nil)
 	if err != nil {
-		klog.Infof("Unable to list buckets, %v\n", err)
+		klog.V(1).Infof("Unable to list buckets, %v", err)
```
```suggestion
		klog.V(1).Errorf("Unable to list buckets, %v", err)
```

kishen-v commented on pkg/client/s3client.go:
```diff
@@ -151,7 +151,7 @@ func (c *S3Client) SelectObjects(bucketName string, regex string) ([]string, err
 	})
 	if err != nil {
-		klog.Infof("failed to list objects, %v\n", err)
+		klog.V(1).Infof("failed to list objects %v", err)
```
```suggestion
		klog.V(1).Errorf("failed to list objects: %v", err)
```

kishen-v commented on pkg/client/s3client.go:
```diff
@@ -187,7 +187,7 @@ func (c *S3Client) CheckIfObjectExists(bucketName, objectName string) (bool, err
 		if aerr, ok := err.(awserr.Error); ok {
 			if aerr.Code() == s3.ErrCodeNoSuchKey {
-				klog.Infof("Object %s not found in %s bucket", objectName, bucketName)
+				klog.V(1).Infof("Object %s not found in %s bucket", objectName, bucketName)
```
```suggestion
				klog.Errorf("object %s not found in %s bucket", objectName, bucketName)
```

KeerthanaAP replied:
Done.

kishen-v commented on cmd/dhcp-sync/dhcp-sync.go:
```diff
@@ -184,10 +184,10 @@ var Cmd = &cobra.Command{
 					if event.Op&fsnotify.Write == fsnotify.Write {
-						klog.Info("modified file:", event.Name)
-						klog.Info("restarting the dhcpd service")
+						klog.V(2).Infof("modified file:%s", event.Name)
+						klog.V(2).Info("restarting the dhcpd service")
```
```suggestion
						klog.V(2).Infof("%s has been modified, proceeding to restart dhcpd service", event.Name)
```

kishen-v reviewed (commented):
There are a few places where the capitalisation is not uniform in the existing code, it can be made uniform across the files, as this PR already covers most lines of logs.
Few other things:
1. Usage of trailing periods. (...)
2. Capitalisation of error statements and formatting them with a : before the format specifier can be looked into to bring uniformity.

kishen-v commented on cmd/image/qcow2ova/qcow2ova.go:
```diff
@@ -159,10 +160,23 @@ Qcow2 images location:
 			}
+			passwordfile, err := os.Create("password.txt")
+			if err != nil {
+				klog.Fatalf("failed to create password file, err: %v", err)
```
Typically after `Fatalf`, os.Exit() is invoked, and hence the `return err` will not be reachable. In order to continue returning the error to the caller, we may use `Errorf`.
Ref: https://pkg.go.dev/k8s.io/klog#Fatalf
```suggestion
				klog.Errorf("failed to create password file, err: %v", err)
```

kishen-v commented on cmd/image/import/import.go:
```diff
@@ -217,11 +217,11 @@ pvsadm image import -n upstream-core-lon04 -b <BUCKETNAME>  --object rhel-83-100
 
 		if !opt.Watch {
-			klog.Infof("Importing Image %s is currently in %s state, Please check the Progress in the IBM Cloud UI\n", *image.Name, *image.State)
+			klog.Infof("Importing Image %s is currently in %s state, Please check the Progress in the IBM Cloud UI", *image.Name, *image.State)
```
```suggestion
			klog.Infof("Image import for %s is currently in %s state, Please check the progress in the IBM cloud UI", *image.Name, *image.State)
```

kishen-v commented on pkg/audit/audit.go:
```diff
@@ -67,15 +67,15 @@ func (a *Audit) Log(name, op, value string) {
 	jsonEntry = append(jsonEntry, '\n')
 	if _, err := a.file.Write(jsonEntry); err != nil {
-		klog.Fatalf("log failed with error: %v", err)
+		klog.Fatalf("log failed with error, err: %v", err)
```
Fatalf may lead to the abrupt termination in case of error. 
We can modify the severity.

KeerthanaAP replied:
Done. Modified the severity if its not expected.

kishen-v commented on test/e2e/sync/sync.go:
```diff
@@ -65,43 +65,43 @@ func runSyncCMD(args ...string) (int, string, string) {
 	}
 
-	klog.Infoln(string(specString))
+	klog.V(3).Info(string(specString))
```
Can add more context: 
```suggestion
	klog.V(3).Info("Specifications for e2e sync test", string(specString))
```

kishen-v commented:
/hold
/lgtm 😃 

cc: @mkumatag

mkumatag commented on cmd/image/qcow2ova/qcow2ova.go:
```diff
@@ -159,10 +160,23 @@ Qcow2 images location:
 			if err != nil {
 				return err
 			}
+			passwordfile, err := os.Create("password.txt")
```
can you please close the file in the defer for safer side?

KeerthanaAP replied:
Done.

mkumatag reviewed (changes requested):
minor one

mkumatag reviewed (approved):
/lgtm

mkumatag commented:
/unhold
